drape_frontend/arrow3d.cpp

```diff
@@ -5,71 +5,365 @@
 
 #include "shaders/program_manager.hpp"
 
+#include "drape/glsl_func.hpp"
+#include "drape/pointers.hpp"
+#include "drape/static_texture.hpp"
 #include "drape/texture_manager.hpp"
 
 #include "indexer/map_style_reader.hpp"
 #include "indexer/scales.hpp"
 
+#include "platform/platform.hpp"
+
+#include "coding/reader.hpp"
+
 #include "geometry/screenbase.hpp"
 
+#include "3party/fast_obj/fast_obj.h"
+
+#include <cstring>
+
 namespace df
 {
-namespace arrow3d {
+namespace arrow3d
+{
 double constexpr kArrowSize = 12.0;
 double constexpr kArrow3dScaleMin = 1.0;
 double constexpr kArrow3dScaleMax = 2.2;
 int constexpr kArrow3dMinZoom = 16;
 }  // namespace arrow3d
 
+namespace
+{
 float constexpr kOutlineScale = 1.2f;
 
-int constexpr kComponentsInVertex = 4;
+int constexpr kComponentsInVertex = 3;
 int constexpr kComponentsInNormal = 3;
+int constexpr kComponentsInTexCoord = 2;
 
 df::ColorConstant const kArrow3DShadowColor = "Arrow3DShadow";
 df::ColorConstant const kArrow3DObsoleteColor = "Arrow3DObsolete";
 df::ColorConstant const kArrow3DColor = "Arrow3D";
 df::ColorConstant const kArrow3DOutlineColor = "Arrow3DOutline";
 
-Arrow3d::Arrow3d(ref_ptr<dp::GraphicsContext> context)
-  : m_arrowMesh(context, dp::MeshObject::DrawPrimitive::Triangles)
-  , m_shadowMesh(context, dp::MeshObject::DrawPrimitive::Triangles)
-  , m_state(CreateRenderState(gpu::Program::Arrow3d, DepthLayer::OverlayLayer))
+std::string const kDefaultArrowMesh = "arrow.obj";
+std::string const kDefaultArrowShadowMesh = "arrow_shadow.obj";
+
+std::string const kMainFileId = "main_obj_file_id";
+
+using TLoadingCompletion = std::function<void(std::vector<float> positions,
+                                              std::vector<float> normals,
+                                              std::vector<float> texCoords)>;
+using TLoadingFailure = std::function<void(std::string const &)>;
+
+namespace fast_obj_adapter
+{
+void * FileOpen(char const * path, void * userData)
+{
+  // Load only the main OBJ file, skip all the files that can be referred
+  // inside the OBJ model.
```

Comment:
```suggestion
  // Load only the main OBJ file, skip all the files that can be referred inside the OBJ model.
```
It fits 120 chars, right?

Reply:
Fixed